tornado/web.py

```diff
@@ -2809,6 +2809,23 @@ def validate_absolute_path(self, root: str, absolute_path: str) -> Optional[str]
                 self.redirect(self.request.path + "/", permanent=True)
                 return None
             absolute_path = os.path.join(absolute_path, self.default_filename)
+        else:
+            canonical_path = os.path.normcase(os.path.normpath(absolute_path))
+            new_url = os.path.relpath(canonical_path, self.root).replace(os.path.sep, '/')
+            if os.path.join(self.root, self.path).replace('/', os.path.sep)\
+                    != canonical_path or not self.request.path.endswith(new_url):
+                if not new_url.startswith('/'):
+                    new_url = '/' + new_url
+                if self.request.path.endswith(new_url):
+                    path_prefix = self.request.path[:-len(new_url)]
+                elif urllib.parse.unquote(self.request.path).endswith(new_url):
+                    path_prefix = urllib.parse.unquote(self.request.path)[:-len(new_url)]
+                else:
+                    raise HTTPError(403, "could not replicate URL parsing for %s", self.path)
+                path_prefix = os.path.normpath(path_prefix).replace(os.path.sep, '/')
+                full_new_url = (path_prefix + new_url).replace(os.path.sep, '/')
```

Comment:
This concatenates `path_prefix`, which is a url path (after the replace call on the previous line), with `new_url`, which I think is already a url path (judging by the name), but then you replace os.path.sep again as if it's a filesystem path.

This feels sloppy - this code needs to have a clear idea of which variables use which rules. Could `pathlib` help?

Reply:
`path_prefix` always starts with a slash, as it is taken from the URL. We should be able to remove the second replace, that's true, as I do it for both `new_url` and `path_prefix` separately already.